src/decode.rs

```diff
@@ -156,11 +155,14 @@ pub fn decode_engine_vec<'e, 'o, E: Engine, T: AsRef<[u8]>>(
     let starting_output_len = buffer.len();
 
     let estimate = engine.decoded_length_estimate(input_bytes.len());
-    let total_len_estimate = estimate
-        .decoded_length_estimate()
+    let len_estimate = estimate.decoded_length_estimate();
+    let total_len_estimate = len_estimate
         .checked_add(starting_output_len)
         .expect("Overflow when calculating output buffer length");
-    buffer.resize(total_len_estimate, 0);
+    buffer.reserve(len_estimate);
+    unsafe {
+        buffer.set_len(total_len_estimate);
+    }
```

Comment:
The safety documentation of [`set_len`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len) is pretty clear that this is UB. Namely it states
> The elements at `old_len..new_len` must be initialized.

But `total_len_estimate = starting_output_len + len_estimate`, which means the last `len_estimate` elements are not properly initialized.

Reply:
@jonasbb It's not UB or unsafe in aggregate since write into that "unintialized" portion of the buffer and then truncate to the true length.

This avoids the overhead of callocing/initializing a buffer that we'll immediately overwrite

Reply:
The documentation mentions that as a safety requirement, and this code is not upholding it. Later, the code also creates a reference to this uninitialized memory, which is also not allowed.
Some functions like `decode` could also panic and unwind *before* truncate could set the correct length. Now, the drop implementation of `Vec` might potentially read the uninitialized values.

If you really need uninitialized memory you need to write to it via a pointer and update the length only after writing. Alternatively, using `MaybeUninit` is an option. If the buffer would be of type `Vec<MaybeUninit<u8>>`, then I think the `set_len` would be ok, but it isn't.